include/eve/detail/function/simd/x86/bitmask.hpp

```diff
@@ -11,74 +11,109 @@
 {
   //================================================================================================
-  // Logical to Bitmap - use movemask variant
+  // Logical to Bits
   //================================================================================================
   template<typename T, typename N, x86_abi ABI>
-  EVE_FORCEINLINE auto to_bitmap( sse2_ const&, logical<wide<T, N, ABI>> const& p ) noexcept
+  EVE_FORCEINLINE auto to_bits( sse2_ const&, logical<wide<T, N, ABI>> const& p ) noexcept
   {
-    using type = std::bitset<N::value>;
+    using type = typename logical<wide<T, N, ABI>>::bits_type;
+    return bit_cast(p.mask(), as_<type>{});
+  }
 
-    if constexpr( std::is_same_v<ABI, x86_128_>)
+  //================================================================================================
+  // Logical to Mask
+  //================================================================================================
+  template<typename T, typename N, x86_abi ABI>
+  EVE_FORCEINLINE wide<T, N, ABI> to_mask(sse2_ const&, logical<wide<T, N, ABI>> const& p ) noexcept
+  {
+    if constexpr( !ABI::regular_logical_register )
     {
-            if constexpr( std::is_same_v<T, float > ) return type(_mm_movemask_ps(p.mask()));
-      else  if constexpr( std::is_same_v<T, double> ) return type(_mm_movemask_pd(p.mask()));
-      else  if constexpr( std::is_integral_v<T> )
-      {
+      auto z = wide<T, N, ABI>(0);
+      auto a = allbits(as_<wide<T, N, ABI>>());
+      auto m = p.storage().value;
 
-              if constexpr( sizeof(T) == 1 ) return type(_mm_movemask_epi8(p.mask()));
-        else  if constexpr( sizeof(T) == 2 )
-        {
-          if constexpr( current_api >= ssse3 )
-          {
-            __m128i mask =  _mm_setr_epi8 ( 0x01,0x03,0x05,0x07,0x09,0x0B,0x0D,0x0F
-                                          , -128,-128,-128,-128,-128,-128,-128,-128
-                                          );
-
-            return type( _mm_movemask_epi8(_mm_shuffle_epi8(p.storage(),mask)) );
-          }
-          else
-          {
-            return type(_mm_movemask_epi8(_mm_packs_epi16(p.storage(), _mm_setzero_si128())));
-          }
-        }
-        else  if constexpr( sizeof(T) >= 4 )
-        {
-          using tgt = logical<wide<as_floating_point_t<T>, N>>;
-          return bit_cast(p,as_<tgt>()).bitmap();
-        }
+      if constexpr( std::same_as<ABI,x86_128_>)
+      {
+              if constexpr( std::same_as<T,double > ) return _mm_mask_blend_pd(m,z,a);
+        else  if constexpr( std::same_as<T,float  > ) return _mm_mask_blend_ps(m,z,a);
+        else  if constexpr( sizeof(T) == 8          ) return _mm_mask_blend_epi64(m,z,a);
+        else  if constexpr( sizeof(T) == 4          ) return _mm_mask_blend_epi32(m,z,a);
+        else  if constexpr( sizeof(T) == 2          ) return _mm_mask_blend_epi16(m,z,a);
+        else  if constexpr( sizeof(T) == 1          ) return _mm_mask_blend_epi8(m,z,a);
+      }
```

Comment:
Why do you need to blend anything - I am confused.
This is not `regular_logical` - so I assume - it's avx512 logical. Why blend?

Maybe I just don't know what `allbits` is though.

Reply:
Because this is`to_mask` which purpose is to turn a logical<wide<T>> into a wide<T>. Not to_bitmap